### ISSUE TITLE
[XCCDF] Check fallback

### ISSUE THREAD

**Issue by balleman-ctr** [usnistgov/SCAP#18]
Presently XCCDF’s check processing picks the first check with a matching selector and a supported system and explicitly does not allow “backtracking” to another available check in the event first cannot be run.  For DISA STIGs, one possibility we have explored is producing SCAP packages that contain combined manual and automated content. If we produce rules with two checks, first OVAL then OCIL, with no selectors, presently SCAP tools should only run the OVAL and never the OCIL. In the event OVAL returns an “error” for instance, the tool should use this as the result, and the OCIL will still not be run.  I would like to suggest changing this behavior so that if a check returns a non-definitive result (error, unknown, notchecked), the tool should execute the next available check (if any) and use its result.  This would allow for falling back to a manual OCIL check (or another system) in the event of an error during an OVAL check.

**Comment by vanderpol** [comment]
@balleman-ctr I think this can be updated just in the XCCDF prose, nothing really needs to change in the XSD rules?   Seems like a straightward idea.

**Comment by balleman-ctr** [comment]
@vanderpol Yes, I think this is only a change in the specification not the schema.  I do not believe we have an editable version of the specification in order to produce an exact proposed change.

**Comment by vanderpol** [comment]
@dragosprisaca we just need to get some of the source material from NIST committed to github so we can determine how to proceed.   Is the XCCDF specification a Word document, or is it generated from some internal system at NIST? 

What should we be doing to help facilitate progress on this?

**Comment by vanderpol** [comment]
@balleman-ctr just something to consider on this is that this scenario may only work in certain circumstances, like when the SCAP scan is scanning the local computer, and likely in a graphical mode.

Using our SCC application as an example, in order to make OCIL less painful, we always prompt the end user for manual questions before any OVAL scan starts, and those answers are cached and applied to all scanning targets at review time.   If SCC is scanning windows or UNIX hosts remotely via WMI or SSH, and some OVAL test ends with a status of 'error', there likely won't be a cached answer for the OCIL questionnaire, as this was overall an 'automated' rule, so we don't make end users answer manual questions when there is an OVAL check that can be used.

If scanning a collection of Windows hosts remotely in our 'classic' mode, we could prompt the user on any OVAL rule with a result of error, but it's going to pause all scanning, and make the process quite a bit less automated.  The same holds true for local scans, but the scope of things being delayed is smaller.

So while I agree at a high level that this could be useful, the devil is in the details on how useful.   For automated scans performed by some service, this likely won't be used at all.  

Even for local scans, end users may not want the scan to pause indefinitely, so we likely would put some type of timer (say 5 minutes) that allows the user to answer the OCIL question, or we just move on with the OVAL error result.

**Comment by vanderpol** [comment]
@balleman-ctr We may actually be able to 'post-scan' prompt the user for rules that error and update the XCCDF results.  Ideally I'd like to capture the fact that the OVAL test errored and then an OCIL question was answered.   

Should this scenario create 2 rule results for the same rule, one that's OVAL and 'error' and one with a later date/timestamp that's OCIL and has the manually answered result?  per the XCCDF schema I think that's valid, but I'm not sure what XCCDF consumers like STIG Viewer and eMass might do with that data, they may need to be updated to know that more than one rule-result for a given rule is possible, and know what to do with them.

**Comment by balleman-ctr** [comment]
@vanderpol – My first thought is that we should only create one rule-result since there is only one outcome of the evaluation.  The details of the OVAL check erroring could be provided in result-details or messages.  I suspect that multiple rule-results in this case would have an undesired impact on scoring.

**Comment by vanderpol** [comment]
@balleman-ctr I think I found the 'correct' place to capture this information:

            <xsd:element name="override" type="cdf:overrideType" minOccurs="0" maxOccurs="unbounded">
                <xsd:annotation>
                    <xsd:documentation xml:lang="en">An XML block explaining how and why an auditor chose to override the result. </xsd:documentation>
                </xsd:annotation>
            </xsd:element>

The overrideType, seems like exactly what we would want to capture.

**Comment by balleman-ctr** [comment]
@vanderpol I'm not certain this would always be an override.  An override suggests a human auditor is disagreeing with the evaluated result.  The check fallback would not necessarily involve a manual input.  You might have, for instance, OVAL and SCE checks for a rule, where if OVAL errors, the SCE check is used.

**Comment by vanderpol** [comment]
@ballemanctr good point, although if OVAL returns an error, normally that's means they system call/api didn't return what was expected (generally not an oval interpreter application error), and I guess SCE might get the data by other means, but it may error as well.   The challenge of 'not always' is tough one, it's really an override all of the time, or not, and 'not always' really means never.  I'm fine with that, we can capture the data in an xccdf info message or something, and this to my view is a pretty rare occurrence.  Unless a system is somehow corrupt, 99+% of oval tests return true or false, error is pretty rare.
